packages/cli/src/cmds/agentInstaller/projectConfiguration.ts

```diff
@@ -98,37 +195,50 @@ async function resolveSelectedInstallers(
  */
 async function getSubprojects(
   dir: string,
-  rootHasInstaller: boolean
+  rootHasInstaller: boolean,
+  availableInstallers: AgentInstaller[]
 ): Promise<ProjectConfiguration[] | undefined> {
-  const ents = await fs.readdir(dir, { withFileTypes: true });
-  const subprojects = (
-    await Promise.all(
-      ents.map(async (ent): Promise<ProjectConfiguration | null> => {
-        if (!ent.isDirectory()) {
-          return null;
-        }
+  let subprojects: ProjectConfiguration[] = [];
+  const yarnSubprojects: ProjectConfiguration[] = await getYarnSubprojects(
+    dir,
+    availableInstallers
+  );
 
-        const subProjectPath = join(dir, ent.name);
-        try {
-          await fs.access(subProjectPath, fsConstants.R_OK);
-        } catch {
-          // This directory is not readable. Ignore it.
-          return null;
-        }
+  if (yarnSubprojects.length > 0) {
+    // detect yarn packages as sub-projects
+    subprojects = yarnSubprojects;
+  } else {
+    // read directory to detect sub-projects
+    const ents = await fs.readdir(dir, { withFileTypes: true });
+    subprojects = (
+      await Promise.all(
+        ents.map(async (ent): Promise<ProjectConfiguration | null> => {
+          if (!ent.isDirectory()) {
+            return null;
+          }
 
-        const installers = await getAvailableInstallers(subProjectPath);
-        if (!installers.length) {
-          return null;
-        }
+          const subProjectPath = join(dir, ent.name);
+          try {
+            await fs.access(subProjectPath, fsConstants.R_OK);
+          } catch {
+            // This directory is not readable. Ignore it.
+            return null;
+          }
 
-        return {
-          name: ent.name,
-          path: subProjectPath,
-          availableInstallers: installers,
-        };
-      })
-    )
-  ).filter(Boolean) as ProjectConfiguration[];
+          const installers = await getAvailableInstallers(subProjectPath);
+          if (!installers.length) {
+            return null;
+          }
+
+          return {
+            name: ent.name,
+            path: subProjectPath,
+            availableInstallers: installers,
+          };
+        })
+      )
+    ).filter(Boolean) as ProjectConfiguration[];
+  }
```

Comment:
Instead of bolting more functionality into this already long function, I'd suggest keeping it as is and using another short function to tie it and `getYarnSubprojects()` together. What do you think?

Reply:
Having an extra short function that calls this one could decrease readability because it adds one more level of indirection the developer must follow to trace this logic.